Dashboard subvenciones

eltercero opened the pull request:
Closes #3076

## :v: What does this PR do?

Add Subsidies Dashboard with the same features Contracts & tenders Dashboard has.

Note: I want to move some files to another folders, but I'm waiting until the previous PRs are done to do so (to avoid conflicts). When it's ready I'll remove the _blocked_ label. 

## :mag: How should this be manually tested?

https://getafe.gobify.net/dashboards/subvenciones/

## :eyes: Screenshots

![Captura de pantalla 2020-05-26 a las 16 16 54](https://user-images.githubusercontent.com/545235/82911680-7d8e8d00-9f6c-11ea-8cf6-e1a4e77d53c9.png)

## :shipit: Does this PR changes any configuration file?

No

## :book: Does this PR require updating the documentation?

No

eltercero committed:
Changing global variables to class attributes

eltercero committed:
Subsidies routes

eltercero committed:
Rails controller, service objects and view changes

eltercero committed:
Preparing GroupPctDistributionBars for long labels

eltercero committed:
Subsidies views and controller (JS)

eltercero committed:
Subsidies tests

eltercero committed:
Adding length to label truncate

ferblape reviewed (changes requested):
![Screenshot 2020-05-26 at 16 49 43](https://user-images.githubusercontent.com/17616/82915368-eb3cb800-9f70-11ea-8f1f-7e1e4ff3db8e.png)

"contratos", copy past is evil 🗡️

ferblape commented on config/locales/gobierto_dashboards/views/ca.yml:
```diff
@@ -44,9 +44,44 @@ ca:
+        date: Data
+        dates: Dates
+        description: Visualització i anàlisi de les subvencions concedides per %{entity_name}
+        fromto: De a Nº Contr.
```
Contr. o Subv?

ferblape commented on config/locales/gobierto_dashboards/views/en.yml:
```diff
@@ -43,9 +42,43 @@ en:
+        date: Date
+        dates: Dates
+        description: Visualizations and analisys for the granted subsidies from %{entity_name}
+        fromto: From To No Contr.
```
Contracts?

ferblape commented on config/locales/gobierto_dashboards/views/es.yml:
```diff
@@ -43,9 +43,43 @@ es:
+        date: Fecha
+        dates: Fechas
+        description: Visualización y análisis de las subvenciones concedidas por %{entity_name}
+        fromto: Desde A Nº. Contr.
```
Contr?

eltercero committed:
Fix bug with truncate function

ferblape commented:
Another issue: you are not limiting the top in the summary and it's showing all the assignees

eltercero committed:
Limiting summary table results

eltercero committed:
Fixing translations

eltercero committed:
One more missing translation

eltercero commented:
@ferblape fixed translations and table limit. I added it to Contracts too, it wasn't there either 🙊

furilo committed:
Markup in Subsidies summary

furilo committed:
Add a nowrap utilitie class

furilo commented:
- [ ] Why there are some quantities with the thousand separators and others not? Happens in both dashboards
![image](https://user-images.githubusercontent.com/24740/82991465-030e4d80-9ffe-11ea-9dd5-905726e3a12f.png)

- [x] In Subvenciones/Subvenciones the quantities columns are not right aligned
![image](https://user-images.githubusercontent.com/24740/82991768-744e0080-9ffe-11ea-842c-c1586c62a1e7.png)

And, fyi, in the Dashboards configuration for the data source with should link to the Dataset, not the Query

eltercero committed:
Fix test.

eltercero committed:
Aligning currency values to the right.

eltercero commented:
>     * [ ]  Why there are some quantities with the thousand separators and others not? Happens in both dashboards

We are using [toLocaleString](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toLocaleString) to handle these (not just in Dashboards) and it seems that this is the way the browser JS engine handles the Spanish locales. It's kind of recent.

![Captura de pantalla 2020-05-27 a las 14 31 52](https://user-images.githubusercontent.com/545235/83019403-2f8a8f80-a027-11ea-866e-1c4fa89ad1f6.png)

More information: https://stackoverflow.com/questions/57628055/tolocalestring-not-working-on-numbers-less-than-10000-in-all-browsers

We could:

1- Find other external libraries to format these
2- Handle the format ourselves in these cases
3- Use locale `de` when formatting currencies in Spanish (both the easiest and ñapest of them)

furilo commented:
OK, let's fork the JS locale issue to an issue.

eltercero committed:
Using a css class instead of an id.

eltercero commented:
@Crashillo ready to review!

eltercero commented:
> OK, let's fork the JS locale issue to an issue.

done #3114

Crashillo commented on app/javascript/gobierto_dashboards/index.js:
```diff
@@ -1,16 +1,33 @@
+});
+
+document.addEventListener('DOMContentLoaded', () => {
+  const subsidiesAppNode = document.getElementById("gobierto-dashboards-subsidies-app");
```
There's no need to declare the same listener twice

eltercero replied:
🤦

Crashillo commented on app/javascript/gobierto_dashboards/modules/subsidies_controller.js:
```diff
@@ -0,0 +1,317 @@
+
+const d3 = { scaleThreshold, sum, mean, median, max }
+
+import * as dc from 'dc'
```
Dead code

Crashillo commented on app/javascript/gobierto_dashboards/modules/subsidies_controller.js:
```diff
@@ -0,0 +1,317 @@
+import { money } from 'lib/shared'
+
+import { AmountDistributionBars } from "lib/visualizations";
+import { GroupPctDistributionBars } from "lib/visualizations";
```
```suggestion
import { AmountDistributionBars, GroupPctDistributionBars } from "lib/visualizations";
```

Crashillo commented on app/javascript/gobierto_dashboards/modules/subsidies_controller.js:
```diff
@@ -0,0 +1,317 @@
+      const subsidy = subsidiesData[i];
+      const amount = subsidy.amount ? parseFloat(subsidy.amount) : 0.0
+
+      let [beneficiary_id, ...beneficiary_name] = subsidy.beneficiary.split(' ')
```
Magic is coming! 👏

Crashillo commented on app/javascript/gobierto_dashboards/modules/subsidies_controller.js:
```diff
@@ -0,0 +1,317 @@
+    const container = this.charts[options.id].container;
+
+    if (options.all) {
+      container.filter(null);
+      container.filter([options.titles]);
```
Explain why is that

Crashillo commented on app/javascript/gobierto_dashboards/modules/subsidies_controller.js:
```diff
@@ -0,0 +1,317 @@
+
+  _formalizedSubsidiesData(subsidiesData){
+    return subsidiesData
+  }
```
I wonder what this do....

Crashillo commented on app/javascript/gobierto_dashboards/webapp/components/DownloadButton.vue:
```diff
@@ -0,0 +1,36 @@
+  methods: {
+    showDownloadDataLink(){
+      return !!this.dataDownloadEndpoint
+    }
```
When is this called?

eltercero replied:
It's not, sorry! It was done between reviews and forgot to remove that one.

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/subsidies_shared/Nav.vue:
```diff
@@ -0,0 +1,63 @@
+        class="dashboards-home-nav--tab"
+        @click.native="markTabAsActive(0)"
+      >
+        <i class="fas fa-chart-bar" />
+        <i class="far fa-chart-bar" />
```
I've seen this. @jorgeatgu, do you remember what happened if we remove the duplicated?

jorgeatgu replied:
Nothing happens, is the same icon but with different weight

far class = font-awesome Regular
fas class = font-awesome Solid

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/subsidy/SubsidiesIndex.vue:
```diff
@@ -0,0 +1,36 @@
+    }
+  },
+  created() {
+    EventBus.$on('refresh_summary_data', () => {
```
$on but no $off

Crashillo commented on app/javascript/gobierto_dashboards/webapp/containers/subsidies_shared/Aside.vue:
```diff
@@ -0,0 +1,168 @@
+    this.initFilterOptions();
+    this.updateCounters(true);
+
+    EventBus.$on('dc_filter_selected', ({title, id}) => {
```
$on but no $off

Crashillo commented on app/javascript/lib/shared/modules/vue-filters.js:
```diff
@@ -27,10 +27,15 @@ export const date = (value, opts = {}) => {
 export const truncate = (value, opts = {}) => {
-  const omission = opts['omission'] || '...';
   const length = opts['length'] || 30;
+  const str = `${value || ''}`
```
Iterating over your commons. That's the way! 🥳

eltercero replied:
I don't understand what you meant, is it ok then 😅 ?

Crashillo replied:
Yes, it's a compliment. :)

Crashillo commented on app/views/gobierto_dashboards/dashboards/subsidies.html.erb:
```diff
@@ -0,0 +1,28 @@
+    </div>
+  </div>
+
+  <div class="js-loading">
```
I think you can add this stuff in the vue app

Crashillo commented on config/locales/gobierto_dashboards/views/ca.yml:
```diff
@@ -44,11 +43,45 @@ ca:
+          label_larger_subsidy_amount: La major subvenció suposa <strong> <span id
+            = "larger-subsidy-amount-pct"> </ span> </ strong> de tota la despesa
+            en subvencions
+          label_less_than_1000: El <strong> <span id = "less-than-1000-pct"> </ span>
```
Back again?

eltercero replied:
Still not sure about the best way to handle these, tbh. Separation is important, but it'll also be messier to handle.

Crashillo replied:
My point of view here is, let's imagine we move the translations to a professional tool, like transifex, why a translator have to know there are HTML in the text and not only this, but even the HTML contains required identificators for the app?

If the translation was: `label_larger_subsidy_amount: La <h1>major</h1> subvenció...`, it wouldn't be OK, but.... mmmm reasonable. But including identificadors? no way.

On the other hand, someone else is developing the controller and need to debug the values. Where are they? OMG, embedded in the translation texts. This is about [comparing apples and oranges](https://www.wordreference.com/es/en/translation.asp?spen=mezclar%20las%20churras%20con%20las%20merinas). Although split the sentence wouldn't be neither the best approach, at least the developer knows where its code is.

eltercero replied:
I did the split, you may check now 👍

Crashillo commented on config/locales/gobierto_dashboards/views/ca.yml:
```diff
@@ -44,11 +43,45 @@ ca:
+          label_less_than_1000: El <strong> <span id = "less-than-1000-pct"> </ span>
+            </ strong> de les subvencions són menors de <strong> 1.000 € </ strong>
+          mean_amount: Import mitjà
+          median_amount: Import mitjà
```
You cannot use the same label for different things, median (maths) is mediano/a

eltercero replied:
I know mean and median are different, but don't they use the same word in catalan? If not, which is it?

ferblape replied:
Yes, they use the same words :D but there are synomis. You can use mitja for mean and mitjana for median

eltercero replied:
👍 I'll change it

Crashillo replied:
https://ca.wikipedia.org/wiki/Mediana

eltercero committed:
Fixes from PR review:

- Changing translations
- Adding comments
- Removing unused code
- Refactor some code
 - Adding .$off to listening events when the component is destroyed

eltercero committed:
Splitting translations with html tags so there are none in the yml files

Crashillo commented on app/javascript/gobierto_dashboards/webapp/components/DownloadButton.vue:
```diff
@@ -0,0 +1,31 @@
+    }
+  },
+  created(){
+    this.isShowDownloadDatalinkShown = !!this.dataDownloadEndpoint
```
When you declare a property with `this`, and then you use it in the DOM, it _should_ be declared in the `data` function.
However, in this case, you can save declaring a new variable for that and simply as:
`v-if="!!this.dataDownloadEndpoint"`

Crashillo commented on config/locales/gobierto_dashboards/views/en.yml:
```diff
@@ -30,24 +29,57 @@ en:
-            id="larger-contract-amount-pct"></span></strong> of all the spending
-          label_less_than_1000: <strong><span id="less-than-1000-pct"></span></strong>
-            of contracts are less than <strong>1.000 €</strong>
+          label_half_spendings_contracts_1: ""
```
Could be another option declare variables? For instance:

yml: `%{value} of contracts accumulate...`
vue: `<p class="description" v-html="I18n.t(LABEL, { value: <strong>VARIABLE</strong> } )">`

Just asking

eltercero replied:
I tried that and it didn't work. Maybe I did something wrong.

Crashillo reviewed (approved):
You can implement the suggestions in the next PR

eltercero committed:
Refactoring variable
